Query Performance

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/21447
## Description

Performance problems could be observed on the global WP list of a QA instance. Turns out the reason is, that the calculation of sums (which takes places even when not showing sums), involves requesting **all** work packages from the database and adding their column values on the rails side.

This PR applies the following fixes to improve the situation:
- for "normal" columns, the aggregation is done SQL side, we only request the sum via SQL
- for custom field columns I couldn't accomplish SQL side aggregation, however we only request work packages where the corresponding custom field has a value
  - this is a huge performance improvement assuming that custom fields are only sparsely filled
  - it might not be an improvement for densely filled custom fields (will probably be worse if there are multiple such summed columns)

Performance-wise improvements are possible (and should be pursuited) in the following areas:
- do not send a SQL query per summed column, but rather use a single SQL request to get all aggregations at once
- move aggregation for custom fields to the SQL side

This PR includes the following additional changes:
- `QueryColumn`s not indicate whether they are summable at all, thus it is not possible to try to sum up "subject" or "project" anymore
  - users are only presented a checkbox for columns that are indeed summable, so they do not try to enable summing for the subject
- summing is now performed inside the `QueryColumn` class, allowing subclasses to override the way sums are calculated
## Related PRs

:warning: Merge this PR only together with the following PRs:
- [Backlogs](https://github.com/finnlabs/openproject-backlogs/issues/169): indicate summable columns
- [Costs](https://github.com/finnlabs/openproject-costs/issues/164): indicate summable columns and fixes n+1 query
## Limitations

At the end I discovered that grouping a query is very slow on its own:
- determining the groups is done ruby-side
- building the group sums involves work package arrays and thus is also slow (can't be performed SQL side)

I decided to not tackle those issues in this PR. I merely added compatibility code to be able to handle the arrays passed by the grouping code. I propose to introduce the changes w.r.t. to grouping in a separate PR, as grouping only impedes the performance when it was explicitly enabled. It will also require further (and probably larger) rework of the query classes to support SQL side grouping.

NobodysNightmare committed:
Always cache work_packages of Query::Results

replacing the already cached :all_work_packages by the usual :work_packages

NobodysNightmare committed:
Move summing of column values into QueryColumn

- letting a column indicate whether it is summable (avoiding errors when trying to sum the "subject" or other silly things)
- summing normal QueryColumns SQL side
- doing SQL side filtering for CF-columns to improve performance for sparsely filled custom fields

Possible improvements:
- fully computing custom field sums SQL side (for dense custom fields)
- only using a single query for all summed columns, instead of one SQL query per column

NobodysNightmare committed:
Only allow to configure columns as summable, that can be summed

no more "sum that subject and the versions!"

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-query-performance.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-query-performance.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
Uniformly format QueryColumns

- wrap all columns parameter-wise
- ensure that none exceeds the 100 character limit per line

NobodysNightmare committed:
Ensure compatibility of sum-functions with Array callers

this is neccessary to allow building sums for grouped results.
The long-term goal should be to also move the grouping to the SQL side.

Currently grouping will make a query very slow
("normal" with respect to its prior performance,
but "slow" with respect to the performance of the grouping-free sum implementation).

ulferts commented on app/models/query_custom_field_column.rb:
```diff
@@ -46,8 +48,24 @@ def custom_field
+                              #{cv_alias}.customized_type='WorkPackage' AND
+                              #{cv_alias}.customized_id=#{WorkPackage.table_name}.id AND
+                              #{cv_alias}.custom_field_id=#{@cf.id}")
+                      .where("#{cv_alias}.value IS NOT NULL AND #{cv_alias}.value != ''")
```
@NobodysNightmare you wanted to double check wether it is possible to use the association here.

NobodysNightmare replied:
I can use it to perform the join, as I don't even need the **outer** join (my `where` will only accept inner join results anyway).

However, the `where` will remain as string, since I believe it is not possible to do these negated statements using AR hashes...

NobodysNightmare committed:
Simplify joins clause

ulferts commented on app/models/query_custom_field_column.rb:
```diff
@@ -46,8 +48,21 @@ def custom_field
+    if work_packages.respond_to?(:joins)
+      # we can't perform the aggregation on the SQL side. Try to filter useless rows to reduce work.
+      work_packages = work_packages
+                      .joins(:custom_values)
```
You are no longer filtering the custom values by the `custom_field_id`

NobodysNightmare replied:
-.-
